Plugins: Angular detector: Remote patterns fetching

xnyo opened the pull request:
<!--

Thank you for sending a pull request! Here are some tips:

1. If this is your first time, please read our contribution guide at https://github.com/grafana/grafana/blob/main/CONTRIBUTING.md

2. Ensure you include and run the appropriate tests as part of your Pull Request.

3. In a new feature or configuration option, an update to the documentation is necessary. Everything related to the documentation is under the docs folder in the root of the repository.

4. If the Pull Request is a work in progress, make use of GitHub's "Draft PR" feature and mark it as such.

5. If you can not merge your Pull Request due to a merge conflict, Rebase it. This gets it in sync with the main branch.

6. Name your PR as "<FeatureArea>: Describe your change", e.g. Alerting: Prevent race condition. If it's a fix or feature relevant for the changelog describe the user impact in the title. The PR title is used to auto-generate the changelog for issues marked with the "add to changelog" label.

7. If your PR content should be added to the What's New document for the next major or minor release, add the **add to what's new** label to your PR. Note that you should add this label to the main PR that introduces the feature; do not add this label to smaller PRs for the feature.

-->

**What is this feature?**

Changes the Angular detection feature so it fetches Angular detection patterns from GCOM, and falls back to the hardcoded ones if there's an error.

The results are cached and re-used until the cache is valid.

This feature is behind a new feature flag: `pluginsRemoteAngularDetectionPatterns`.

Currently, the patterns are only fetched when testing for Angular plugins (upon startup and when installing new plugins), when the cached result has expired.

This PR does not periodically refresh the pattern and updates the angular status at runtime. Once a plugin has been tested for angular, it won't be checked again if the remote patterns change.

The changes in GCOM haven't been merged yet. Until then, if you want to test this locally, you need a local instance of GCOM on this PR's branch:https://github.com/grafana/grafana-com/pull/6723

And you have to change grafana's config with:

```ini
[grafana_net]
url = http://127.0.0.1:4000
```

And enable the feature flag:

```ini
[feature_toggles]
enable = pluginsRemoteAngularDetectionPatterns
```

**Why do we need this feature?**

Angular deprecation.

**Who is this feature for?**

Grafana users.

**Which issue(s) does this PR fix?**:

<!--

- Automatically closes linked issue when the Pull Request is merged.

Usage: "Fixes #<issue number>", or "Fixes (paste link of issue)"

-->

Related to #68974 

**Special notes for your reviewer:**

- Requires changes in GCOM: https://github.com/grafana/grafana-com/pull/6723
- Enterprise PR: https://github.com/grafana/grafana-enterprise/pull/5269

Please check that:
- [x] It works as expected from a user's perspective.
- [x] If this is a pre-GA feature, it is behind a feature toggle.
- [ ] The docs are updated, and if this is a [notable improvement](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/#how-to-determine-if-content-belongs-in-a-whats-new-document), it's added to our [What's New](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/) doc.

xnyo committed:
Plugins: Angular detector: Remote patterns fetching

xnyo committed:
Renamed PatternType to GCOMPatternType

xnyo committed:
Renamed files

xnyo committed:
Renamed more files

xnyo committed:
Moved files again

xnyo committed:
Add type checks, unexport GCOM structs

xnyo committed:
Cache failures, update log messages, fix GCOM URL

xnyo committed:
Fail silently for unknown pattern types, update docstrings

xnyo committed:
Fix tests

xnyo committed:
Rename gcomPattern.Value to gcomPattern.Pattern

xnyo committed:
Refactoring

xnyo committed:
Add FlagPluginsRemoteAngularDetectionPatterns feature flag

xnyo committed:
Merge branch 'main' into giuseppe/angular-deprecation/remote-patterns

xnyo committed:
Fix tests

xnyo committed:
Merge branch 'main' into giuseppe/angular-deprecation/remote-patterns

xnyo committed:
Merge branch 'main' into giuseppe/angular-deprecation/remote-patterns

xnyo committed:
Re-generate feature flags

xnyo committed:
Add TestProvideInspector, renamed TestDefaultStaticDetectorsInspector

xnyo committed:
Add TestProvideInspector

xnyo committed:
Add TestContainsBytesDetector and TestRegexDetector

xnyo committed:
Renamed getter to provider

xnyo committed:
More tests

xnyo committed:
TestStaticDetectorsProvider, TestSequenceDetectorsProvider

xnyo committed:
GCOM tests

xnyo committed:
Lint

xnyo committed:
Merge branch 'main' into giuseppe/angular-deprecation/remote-patterns

xnyo committed:
Made detector.detect unexported, updated docstrings

xnyo committed:
Allow changing grafana.com URL

xnyo committed:
Merge branch 'main' into giuseppe/angular-deprecation/remote-patterns

xnyo committed:
Fix API path, add more logs

xnyo committed:
Update tryUpdateRemoteDetectors docstring

andresmgot commented on pkg/plugins/manager/loader/angulardetector/angulardetector.go:
```diff
@@ -1,46 +1,60 @@
-	return d.regex.Match(moduleJs)
+// newHardcodedInspector returns the default Inspector, which is a PatternsListInspector that only uses the
+// hardcoded (static) angular detection patterns.
+func newHardcodedInspector() (Inspector, error) {
```
since we are using the same approach than when retrieving the public manifest key, it would be nice to follow the same approach. In that case:

 - Define the interface that both the hardcoded and the remote inspectors will follow in the `plugins` package (`ifaces.go`). BTW, there we use  `static` for what you can `hardcoded` and `dynamic` for what you call `remote`. Probably it's better to use the same naming.
 - Here within the plugin manager, you expose the default behavior (static/hardcoded). You can see `static_retriever.go` for how it's implemented there.
 - Then, in `pluginsintegrations` you define the dynamic/remote version of the service and the wrapper that instantiates the static or the dynamic version based on configuration. The existing package there that does this is called `keyretriever`.

xnyo replied:
:+1:  sounds good, I will refactor it so it's similar to signing keys retrieval

xnyo replied:
This will go into a follow up PR, so I'm marking this one as resolved

andresmgot commented on pkg/plugins/manager/loader/angulardetector/gcom.go:
```diff
@@ -0,0 +1,204 @@
+	p.mux.Lock()
+	defer p.mux.Unlock()
+
+	if err := p.tryUpdateRemoteDetectors(ctx); err != nil {
```
If I am reading this correctly, this will block the startup process until the detectors are fetched (if the entries are empty or no longer valid). We should try to avoid that if possible. The approach we followed for the public manifest was to make this a background service, and eventually update the value (in this case, the detectors). In the first run, the dynamic detector would return the hardcoded detectors but in the next server start, it will use the cached values that were fetched asynchronously. 

Note that you need to use persistent storage for that so look at how that's done for the public manifest.

xnyo replied:
Correct, I did it this way as I didn't want to use the storage, but I can refactor it so they are similar and use the storage like for the signing keys

xnyo replied:
I will remove caching from this PR as suggested, so resolving this

andresmgot commented on pkg/plugins/manager/loader/angulardetector/inspector.go:
```diff
@@ -0,0 +1,45 @@
+// Inspector can inspect a plugin and determine if it's an Angular plugin or not.
+type Inspector interface {
+	// Inspect takes a plugin and checks if the plugin is using Angular.
+	Inspect(ctx context.Context, p *plugins.Plugin) (bool, error)
```
can you use different packages for the "inspector" and the "detector"? it will help for the readability and the separation of concerns.

xnyo replied:
:+1: I agree it makes sense to separate those

andresmgot reviewed (commented):
thanks for the PR! I think this falls into the same pattern than the public manifest key retrieval so I have some comments so that we follow the same approach (and avoid the same problems). 

BTW, it's better if you don't try to fit everything in a single PR, it's difficult to review changes that way. A suggestion for a possible split (that is merge-friendly):

 - PR 1: Define the new feature flag. Return an error if enabled.
 - PR 2: Define interfaces and the "hardcoded" implementation of it. This can maybe be included with the first PR.
 - PR 3: Implementation for the "remote" implementation within the feature flag.
 - PR 4: Cache implementation.
 
Let me know what you think :+1:

marefr commented on pkg/plugins/manager/loader/angulardetector/gcom.go:
```diff
@@ -0,0 +1,204 @@
+	if err != nil {
+		return nil, fmt.Errorf("new request with context: %w", err)
+	}
+	resp, err := http.DefaultClient.Do(req)
```
Nit. Not using  the HTTP client created in `newGCOMDetectorsProvider`

marefr commented on pkg/plugins/manager/loader/angulardetector/inspector.go:
```diff
@@ -0,0 +1,45 @@
+
+func (i *PatternsListInspector) Inspect(ctx context.Context, p *plugins.Plugin) (isAngular bool, err error) {
+	f, err := p.FS.Open("module.js")
+	if err != nil {
```
Nit not sure who calls this and what it does with the error. Maybe return false without an error if module.js doesn't exist - there are some backend plugins that might not have a module.js?

marefr commented on pkg/plugins/manager/loader/loader.go:
```diff
@@ -185,7 +186,11 @@ func (l *Loader) loadPlugins(ctx context.Context, src plugins.PluginSource, foun
 			var err error
-			p.AngularDetected, err = l.angularInspector.Inspect(p)
+
+			cctx, canc := context.WithTimeout(ctx, time.Minute*1)
```
1 minute sounds high given that this happens when Grafana starts. Could it really take that long time to run the detection for one plugin?

xnyo replied:
Agree, I was planning to take a better look at tweaking the timeout/refactor to background service similar to the signing keys retrieval after merging this initial PR. Currently it is under a feature flag, so it's safe to merge, but I can the refactoring into background service it directly in this PR as suggested by Andres. I believe once I do that, the context arg could go away (as the GCOM call will happen in the background) so it should solve this concern as well :smiley:

marefr replied:
Make sense :+1:

marefr commented on pkg/services/pluginsintegration/pluginsintegration.go:
```diff
@@ -52,7 +52,7 @@ var WireSet = wire.NewSet(
 	pluginscdn.ProvideService,
 	assetpath.ProvideService,
-	angulardetector.ProvideService,
+	angulardetector.ProvideInspector,
```
Related to @andresmgot comment about splitting up detection and inspection in separate packages becomes clear  why that would be a good idea with this example.

marefr reviewed (commented):
Looks good. Added some comments, but in general agree with @andresmgot earlier review.

xnyo committed:
Use angulardetector http client

xnyo committed:
Return false, nil if module.js does not exist

xnyo committed:
Chore: Split angualrdetector into angularinspector and angulardetector packages

Moved files around, changed references and fixed tests:
- Split the old angulardetector package into angular/angulardetector and angular/angularinspector
- angulardetector provides the detection structs/interfaces (Detector, DetectorsProvider...)
- angularinspector provides the actual angular detection service used directly in pluginsintegration
- Exported most of the stuff that was private and now put into angulardetector, as it is not required by angularinspector

xnyo committed:
Renamed detector.go -> angulardetector.go and inspector.go -> angularinspector.go

Forgot to rename those two files to match the package's names

xnyo committed:
Renamed angularinspector.ProvideInspector to angularinspector.ProvideService

xnyo committed:
Renamed "harcoded" to "static" and "remote" to "dynamic"

from PR review, matches the same naming schema used for signing keys fetching

marefr reviewed (approved):
LGTM given this is behind feature toggle and we can continue work in follow up PRs. Great work

xnyo committed:
Merge branch 'main' into giuseppe/angular-deprecation/remote-patterns

xnyo committed:
Fix merge conflict on updated angular patterns

andresmgot commented on pkg/plugins/manager/loader/angular/angulardetector/gcom.go:
```diff
@@ -0,0 +1,204 @@
+		d, err := pattern.detector()
+		if err != nil {
+			// Fail silently in case of an errUnknownPatternType.
+			// This allows us to introduce new pattern types without breaking old Grafana versions
```
should at least debug the error?

xnyo replied:
Yep good idea, however it's a bit hard to do now because we have no logger in there current. I can leave it for a folow-up PR maybe?

andresmgot replied:
Up to you because it's so small you may forget. It should be easy to add, just make a struct arround gcomPatterns and forward the logger from GCOMDetectorsProvider.

andresmgot commented on pkg/plugins/manager/loader/angular/angulardetector/angulardetector.go:
```diff
@@ -0,0 +1,69 @@
+// Detector implements a check to see if a js file is using angular APIs.
+type Detector interface {
+	// Detect takes the content of a js file and returns true if the plugin is using Angular.
+	Detect(js []byte) bool
```
nit: this is later used as `detector.Detect(...)` which doesn't say what we are detecting. I suggest to change it to `DetectAngularPatterns` or something like that.

andresmgot commented on pkg/services/pluginsintegration/pluginsintegration.go:
```diff
@@ -52,7 +52,7 @@ var WireSet = wire.NewSet(
 	pluginscdn.ProvideService,
 	assetpath.ProvideService,
-	angulardetector.ProvideService,
+	angularinspector.ProvideService,
```
I would move the service definition to this package. The static detector can stay in the plugins manager as an exposed function. (We may want to change the inspector implementation in other scenarios?)

andresmgot reviewed (commented):
Just some minor comments :+1:

xnyo committed:
Removed GCOM cache

xnyo committed:
Renamed Detect to DetectAngular and Detector to AngularDetector

xnyo committed:
Merge remote-tracking branch 'origin/giuseppe/angular-deprecation/remote-patterns' into giuseppe/angular-deprecation/remote-patterns

xnyo committed:
Fix call to NewGCOMDetectorsProvider in newDynamicInspector

xnyo committed:
Removed unused test function newError500GCOMScenario

xnyo commented:
@torkelo Right now it's not a problem, but we are planning to show users which panels/datasources in a dashboard are affected by Angular deprecation (without having to go into "edit" mode), to make it easier for the user to identify what will stop working and needs to be migrated. If we don't do this, some users may end up with broken dashboards once we remove Angular entirely, and they may not have been aware of the deprecation.

The patterns we have now are quite accurate, but we have had some false positives/negatives historically from time to time.

If the patterns are wrong, we risk annoying the user with false warnings, or they may miss those warnings entirely if we have a missing pattern. By having the angular patterns remotely, we can tweak the detections for existing instances (mostly for non-HG users, like on-prem) without forcing the user to update to a new version of Grafana :smile:

xnyo commented on pkg/plugins/manager/loader/angular/angulardetector/gcom.go:
```diff
@@ -0,0 +1,204 @@
+// baseURL is the GCOM base url, without /api and without a trailing slash (e.g.: https://grafana.com)
+// A default reasonable value for ttl is defaultGCOMDetectorsProviderTTL.
+func NewGCOMDetectorsProvider(baseURL string, ttl time.Duration) (DetectorsProvider, error) {
+	cl, err := httpclient.New()
```
Good point :+1: , I am removing caching entirely from this PR as it's behind a feature flag anyways, so I am resolving this for now

torkelo commented:
@xnyo makes some sense :)  but updating the patterns in patch release seems like it would have been a workable solution as well :)

xnyo committed:
Added angularinspector service definition in pluginsintegration

xnyo committed:
Merge branch 'main' into giuseppe/angular-deprecation/remote-patterns

xnyo commented:
@andresmgot I have removed caching entirely in this PR, renamed the interface and moved the inspector service definition to pluginsintegration as suggested :smile: Please can you take a look now? It's probably easier to review by viewing the commits I made today rather than the diff as I have mostly renamed/(re)moved some things without changing implementations

andresmgot commented on pkg/services/pluginsintegration/angularinspector/angularinspector.go:
```diff
@@ -0,0 +1,25 @@
+	var underlying angularinspector.Inspector
+	var err error
+	if cfg.Features != nil && cfg.Features.IsEnabled(featuremgmt.FlagPluginsDynamicAngularDetectionPatterns) {
+		underlying, err = angularinspector.NewDynamicInspector(cfg)
```
you are instantiating the service here in `pluginsintegration` but the code is in `pkg/plugins/manager` (and it's only imported here). My suggestion is to leave only the interface in  `pkg/plugins` and move the implementation code here.

xnyo replied:
I am a bit confused by this and what should live in `pkg/plugins` vs `pluginsintegration`.

The angular inspector itself [doesn't require anything outside of `pkg/plugins`](https://github.com/grafana/grafana/blob/18a6fa861831337bf0571ccb920b42df0fb9d6e3/pkg/plugins/manager/loader/angular/angularinspector/angularinspector.go#L10), however we need `pkg/services/featuremgmt` to switch between the detectors provider depending on feature flag (which is defined in `pkg/services/featuremgmt`), so I have moved the code that switches the implementation in here, and kept the generic inspector implementation in `pkg/plugins` because all its dependencies are in `pkg/plugins`.

I was under the impression that `pluginsintegration` was only for code that depends on other packages outside of `pkg/plugins` (like if it requires kvstore, featuremgmt, pkg/infra, or other Grafana-specific services).

If so, I would prefer leaving the default inspector inside `pkg/plugins`, as the angular detector is required by the plugins loader, and its default implementation also lives in `pkg/plugins`.

@marefr wdyt?

---

I have also noticed that we have [`Features plugins.FeatureToggles` defined in the plugin-specific `Cfg` struct](https://github.com/grafana/grafana/blob/main/pkg/plugins/config/config.go#L41) (it looks like it's unused), so technically it could be used to test for feature flags:

```go
cfg.Features.IsEnabled(featuremgmt.FlagPluginsDynamicAngularDetectionPatterns)
//                     ^~~~~~~~~~~
```

The problem is that we may need to re-define the feature flag if we need to use it `pkg/plugins`, because `featuremgmt` (where the generated feature flags consts are defined) is outside of `pkg/plugins`, and we try to avoid referencing other Grafana packages in `pkg/plugins`.

andresmgot replied:
> I am a bit confused by this and what should live in `pkg/plugins` vs `pluginsintegration`.

yeah, I also find it a bit confusing :disappointed: 

> I was under the impression that `pluginsintegration` was only for code that depends on other packages outside of `pkg/plugins` (like if it requires kvstore, featuremgmt, pkg/infra, or other Grafana-specific services).

What I understood is that services `pluginsintegrations` are suitable to be implemented in a different way for other Grafana environments. In theory, that means that we could redefine the inspector to do something else rather than using gcom to retrieve the patterns (or do nothing).

Again, as I understood, in `pkg/plugins` we can define the default behavior. In this case, that would be returning the hardcoded list. Then, in the `pluginsintegrations` we define the service that "optionally" download new patterns dynamically.

As far as Grafana core is concerned, _something_ should implement the inspector interface but the specifics don't need to live in core. Is up to wire to resolve that dependency.

xnyo replied:
Makes sense, I have moved all the gcom-related features (inspector and also patterns provider) to pluginsintegration :smiley:

andresmgot reviewed (commented):
just a couple clarifications more :+1:

xnyo committed:
Moved dynamic inspector into pluginsintegration

xnyo committed:
Move gcom angulardetectorsprovider into pluginsintegration

xnyo committed:
Log errUnknownPatternType at debug level

xnyo committed:
merge with main

xnyo committed:
re-generate feature flags

xnyo committed:
fix error log

andresmgot reviewed (approved):
Cool :+1: Thanks for the changes!

xnyo committed:
Merge branch 'main' into giuseppe/angular-deprecation/remote-patterns
